### ISSUE TITLE
[DesignProposal]: train operator capacity

### ISSUE THREAD

**Issue by zc277584121** [towhee-io/towhee#453]
### Is there an existing issue for this?

- [x] I have searched the existing issues

### What would you like to be added?

# Background

Currently, there are more and more operators and pipelines in towhee hub, which are build with neural network.
But now we can not train them in the towhee framework.
In machine learning, training is a very important capacity, with which we can easily fine tune a big neural network in a short time with a pretrained model.
So we need to develop the towhee training interface for users.
First, we need to make the operator trainable. And after this, we will design the pipeline training capacity.

# Overview

We need to provide users with an easy-to-use operator training interface.
It needs to adapt to various of frameworks, datasets, all levels training, popular network models.

## training module checklist

- [X] survey, disscuss and framework design.
- [x] operator training framework constructing.
- [x] build a trainable cnn operator and test it.
- [x] training config module constructing.
- [x] basic method implement about training.
- [x] metrics module.
- [x] callback module.
- [x] loss, optimizer and learning rate.
- [x] dataset module.
- [x] log and verbosity.
- [x] more network categories test.
- [ ] more package integration.
- [ ] visualization

# Fine-tuning an operator design
## Quick start
You can fine-tune an operator in only just 3 lines of code:
```python
from towhee.dataset import get_dataset
from towhee.trainer import Trainer, TrainingConfig

op = op('resnet-image-embedding', model_name='resnet101')
mnist_dataset = get_dataset('mnist')
op.train(train_dataset=mnist_train_data)
```

When you run this code, you will see the training process information printed on the screen during every epoch.
In fact, you can train an operator in only 3 steps:
### Step1. Get the instance of operator 
Your must prepare an operator before training. Make sure that there is an operator class code locally and you can construct its instance. The operator class must extend from NNoperator.
```python
from [yourcode] import EfficientnetImageEmbedding
# EfficientnetImageEmbedding is an operator extend from NNOpertor
op = EfficientnetImageEmbedding('efficientnet-b5')
```
Or, you can download an operator class code from towhee hub by name.
```python
op = towhee.op('resnet-image-embedding', model_name='resnet101')
```
### Step2. Get the dataset
You can use the method get_dataset to get a dataset easily by passing the name param.
```python
mnist_dataset = towhee.dataset('mnist')
```

For more parameters you can get from dataset.
### Step3. Train it.
```python
op.train(train_dataset=mnist_train_data)
```

If you start training without setting the config, the operator will get a set of configs by default.
Also, you can specify a config and pass it to the train method.
```python
training_config = TrainingConfig(
    output_dir='./temp_output',
    epoch_num=2,
    batch_size=8,
    prediction_loss_only=True,
)
op.train(training_config, train_dataset=mnist_train_data)
```
The default config is in an yaml file, and you can define your custom config in a new yaml file and just load it to the training_config.
```python
training_config = TrainingConfig(config_file='your_yaml.yaml')
op.train(training_config, train_dataset=mnist_train_data)
```
Your yaml file can be defined like this
```yaml
output_dir: ./output
batch_size: 16
epoch_num: 5
max_steps: -1
loss: BCE
optimizer: Adam
lr_scheduler_type: linear
warmup_ratio: 0
warmup_steps: 0
overwrite_output_dir: True
prediction_loss_only: False
```
More parameters in config you can see training config.
Save operator model weights
Manually, you can save model's weights after training.
```python
op.save_weights('yolov5-v1')
```

### Load model from pretrained weights
You can load weights from a pre-trained weights file and then train the model.
```python
op.load_weights('weights.pt')
op.train()
```
Push to hub
```python
op.push_to_hub(...)
```

### Show the model card
The method op.show_model_card() can print the trained model information formattly
```python
op.show_model_card()
#######################################
#---------model card--------
# model name: resnet101
# task: classification
# training dataset: imagenet
# ....
```

### Define your custom train step
More flexible, you can define your own training step behavior in the training loop.
```python
class MyTrainer(Trainer):
    def train_step(self, batch, batch_index)
        # your custom train step in one batch iter
op.trainer = MyTrainer()
op.train()
```
